pip/pip-278.md

```diff
@@ -0,0 +1,158 @@
+![pip-278-1.png](images/pip-278-1.png)
+
+# Detailed Design
+
+## Design & Implementation Details
+
+* Define a standard TopicCompactionService interface.
+
+  ```java
+  import javax.annotation.Nonnull;
+  
+  public interface TopicCompactionService {
+      /**
+       * Compact the topic.
+       * Topic Compaction is a key-based retention mechanism. it will keep the most recent value for a given key and
+       * user will read compacted data from TopicCompactionService.
+       *
+       * @return a future that will be completed when the compaction is done.
+       */
+      CompletableFuture<Void> compact();
+
+      /**
+       * Read the compacted entries from the TopicCompactionService.
+       *
+       * @param startPosition         the position to start reading from.
+       * @param numberOfEntriesToRead the maximum number of entries to read.
+       * @return a future that will be completed with the list of entries, this list can be null.
+       */
+      CompletableFuture<List<Entry>> readCompactedEntries(@Nonnull Position startPosition, int numberOfEntriesToRead);
+
+      /**
+       * Read the last compacted entry from the TopicCompactionService.
+       *
+       * @return a future that will be completed with the compacted last entry, this entry can be null.
+       */
+      CompletableFuture<Entry> readLastCompactedEntry();
+
+      /**
+       * Get the last compacted position from the TopicCompactionService.
+       *
+       * @return a future that will be completed with the last compacted position, this position can be null.
+       */
+      CompletableFuture<Position> getLastCompactedPosition();
+  }
+  ```
+
+* Define a standard CompactionServiceFactory interface to manage `TopicCompactionService`.
+
+  ```java
+  public interface CompactionServiceFactory extends AutoCloseable {
+
+    /**
+     * Initialize the compaction service factory.
+     *
+     * @param pulsarService
+     *            the pulsar service instance
+     * @return a future represents the initialization result
+     */
+    CompletableFuture<Void> initialize(PulsarService pulsarService);
+
+    /**
+     * Create a new topic compaction service for topic.
+     *
+     * @param topic
+     *            the topic name
+     * @return a future represents the topic compaction service
+     */
+    CompletableFuture<TopicCompactionService> newTopicCompactionService(String topic);
+  }
+  ```
+
+* Implement `PulsarCompactionServiceFactory` and `PulsarCompactionService`
+
+* Combining `CompactedTopicImpl` and `TwoPhaseCompactor` to `PulsarTopicCompactionService`
+
+* Rename `CompactorSubscription` to `PulsarCompactorSubscription`, since it is only applicable to the implementation of Pulsar.
+
+* For CompactorMetrics keep the current implementation, and it only supports `PulsarTopicCompactionService` currently, in the future will use the `Otel` API or other metrics API instead, and customize TopicCompactedService should implement it.
```

Comment:
```suggestion
* For `CompactorMetrics`: keep the current implementation. Currently, it only supports `PulsarTopicCompactionService`. In the future, it will use the `Otel` API or other metrics API instead, and customize TopicCompactedService should implement it.
```
1. Do you mean this?

2. What do you intend to convey? "and customize TopicCompactedService should implement it."

Reply:
> What do you intend to convey? "and customize TopicCompactedService should implement it."

Customized `TopicCompactedService` should implement the `Otel` API or other metrics API.

Reply:
Then that sentence can be:
 "In the future, `CompactorMetrics` will use the `Otel` API or other metrics API. Customized `TopicCompactedService` should implement the Otel API or other metrics API."

It's better to show the objects explicitly rather than using a fuzzy word like `it`. Users do not what does "it" refer to.

Reply:
Thanks, I improve it.